### PR TITLE
Add code to handle pagination of parts. Fixes max layer size of 10GB bug

### PR DESCRIPTION
**Pull request by bainsy88**

### CODE REVIEW

File: registry/storage/driver/s3-aws/s3.go [hunk 1]
```diff
@@ -587,11 +587,20 @@ func (d *driver) Writer(ctx context.Context, path string, append bool) (storaged
 		if err != nil {
 			return nil, parseError(path, err)
 		}
-		var multiSize int64
```
Comment (bainsy88) [inline registry/storage/driver/s3-aws/s3.go]: This multiSize was never used so is redundant code.  The size is calculated by the calling function

File: registry/storage/driver/s3-aws/s3.go [hunk 2]
```diff
@@ -549,9 +549,9 @@ func (d *driver) Reader(ctx context.Context, path string, offset int64) (io.Read
 
 // Writer returns a FileWriter which will store the content written to it
 // at the location designated by "path" after the call to Commit.
-func (d *driver) Writer(ctx context.Context, path string, append bool) (storagedriver.FileWriter, error) {
```
Comment (bainsy88) [inline registry/storage/driver/s3-aws/s3.go]: Setting a variable with the name `append` overwrites the built-in function which meant I couldn't use it further down.

**Comment by bainsy88** [comment]
Yeah, agree on testing. I have manually tested for what it's worth. The reason for no unit tests up front is just time constraints, looks like there isn't a framework for mocking out S3 calls at the moment, so would be a fair chunk of work to get that implemented. The S3 interface is huge as well which doesn't make it very easy!

**Comment by beanssoft** [comment]
Hi guys! is there a time frame to get this done? I got an use case that precisely needs of this feature enabled, thanks!

File: registry/storage/driver/s3-aws/s3.go [hunk 3]
```diff
@@ -587,11 +587,20 @@ func (d *driver) Writer(ctx context.Context, path string, append bool) (storaged
 		if err != nil {
 			return nil, parseError(path, err)
 		}
-		var multiSize int64
-		for _, part := range resp.Parts {
-			multiSize += *part.Size
+		allParts = append(allParts, resp.Parts...)
+		for *resp.IsTruncated {
+			resp, err = d.S3.ListParts(&s3.ListPartsInput{
```
Comment (jinuxstyle) [inline registry/storage/driver/s3-aws/s3.go:592]: The ListParts only list the parts ranging from 1000 to 2000, right? What if the number of parts is more than 2000?

**Review by arkodg — APPROVED** [review]
LGTM, @bainsy88
can you please run `make test` and set these env vars https://github.com/docker/distribution/blob/f4506b517a7e9e64cfdcc73c991d9e85b784dc3e/registry/storage/driver/s3-aws/s3_test.go#L107 to make sure the S3 tests run

**Comment by milosgajdos** [comment]
@bainsy88 is there some public image I can test this on? Would be good to get this merged in.

**Comment by milosgajdos** [comment]
ping @bainsy88 :)

**Review by milosgajdos — APPROVED** [review]

**Review by dmcgowan — APPROVED** [review]
LGTM

**Comment by oingooing** [comment]
hi,
It has been merged, but it seems that it has not been released yet.
I want to patch self, how can I do that?
Do you have any patch files?

**Pull request merged**